src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -415,10 +438,43 @@ internal SynthesizedAttributeData SynthesizeDebuggerStepThroughAttribute()
             return TrySynthesizeAttribute(WellKnownMember.System_Diagnostics_DebuggerStepThroughAttribute__ctor);
         }
 
-        internal SynthesizedAttributeData SynthesizeReadOnlyAttribute()
+        internal void EnsureIsReadOnlyAttributeExists(DiagnosticBag diagnostics, Location location, bool modifyCompilation)
         {
-            // PROTOTYPE(readonlyRefs) it is optional now as it will be generated in the next PR
-            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_ReadOnlyAttribute__ctor, isOptionalUse: true);
+            var isNeeded = CheckIfIsReadOnlyAttributeNeeded(diagnostics, location);
+
+            if (isNeeded && modifyCompilation)
+            {
+                NeedsGeneratedIsReadOnlyAttribute = true;
+            }
+        }
+
+        internal bool CheckIfIsReadOnlyAttributeNeeded(DiagnosticBag diagnosticsOpt, Location locationOpt)
```

Comment:
>CheckIfIsReadOnlyAttributeNeeded [](start = 22, length = 32)

It feels like the name is somewhat confusing. We know the attribute is needed, what we want to know if we need to embed it. Consider renaming to CheckIfIsReadOnlyAttributeSholdBeEmbedded.